rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+    unreachable_pub,
+    unused_qualifications
+)]
+
+// Most of the code in this file is generated, so sometimes it generates code
+// that doesn't follow the clippy rules. We don't want to see those warnings.
+#[allow(clippy::too_many_lines, clippy::use_self)]
+mod bindings {
+    // In `build.rs`, we generate bindings based on the config.yml file. Here is
+    // where we pull in those bindings and make them part of our library.
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+use std::ffi::c_char;
+use std::{ffi::CStr, marker::PhantomData};
+use std::mem::MaybeUninit;
+use std::ptr::NonNull;
+
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+pub use self::bindings::*;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Debug)]
+pub struct Diagnostic<'pr> {
+    diagnostic: NonNull<yp_diagnostic_t>,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Diagnostic<'pr> {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    ///
+    /// Panics if the message is not able to be converted into a `CStr`.
+    ///
+    #[must_use]
+    pub fn message(&self) -> &str {
+        unsafe {
+            let message: *mut c_char = self.diagnostic.as_ref().message as *mut c_char;
+            CStr::from_ptr(message).to_str().expect("YARP allows only UTF-8 for diagnostics.")
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Debug)]
+pub struct Comment<'pr> {
+    comment: NonNull<yp_comment_t>,
+    marker: PhantomData<&'pr yp_comment_t>
+}
+
+impl<'pr> Comment<'pr> {
+    /// Returns the text of the comment.
+    ///
+    /// # Panics
+    /// 
+    /// Panics if the comment is not able to be read.
+    /// 
+    #[must_use]
+    pub fn text(&self) -> &[u8] {
+        unsafe {
+            let start = self.comment.as_ref().start.cast::<u8>();
+            let end = self.comment.as_ref().end.cast::<u8>();
+
+            let len = usize::try_from(end.offset_from(start)).expect("end should point to memory after start");
+            std::slice::from_raw_parts(start, len)
+        }
+    }
+}
+
+/// A struct created by the `errors` or `warnings` methods on `ParseResult`. It
+/// can be used to iterate over the diagnostics in the parse result.
+pub struct Diagnostics<'pr> {
+    diagnostic: *mut yp_diagnostic_t,
+    marker: PhantomData<&'pr yp_diagnostic_t>
+}
+
+impl<'pr> Iterator for Diagnostics<'pr> {
+    type Item = Diagnostic<'pr>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.diagnostic.is_null() {
+            None
+        } else {
+            let current = Diagnostic {
+                diagnostic: unsafe { NonNull::new_unchecked(self.diagnostic) },
+                marker: PhantomData
+            };
+
+            self.diagnostic = unsafe { (*(self.diagnostic)).node.next.cast::<yp_diagnostic_t>() };
+            Some(current)
+        }
```

Comment:
this can be made more maintainable and use less unsafe doing this:

```suggestion
        if let Some(diagnostic) = NonNull::new(self.diagnostic) {
            let current = Diagnostic {
                diagnostic,
                marker: PhantomData
            };

            self.diagnostic = unsafe { (*(self.diagnostic)).node.next.cast::<yp_diagnostic_t>() };
            Some(current)
        } else {
            None
        }
```

Reply:
it looks like this is similar to the `impl Iterator for Comment` below. It'd be good to standardize across these impls.

I prefer the `if let` style vs `match` since it removes one layer of rightward drift, but both are equivalent.